### ISSUE TITLE
Modify no-restricted-globals to allow providing optional custom messages

### ISSUE THREAD

**Issue by dryajov** [eslint/eslint#8315]
@not-an-aardvark @platinumazure

This is a continuation of a new rule proposal and subsequent discussion initially started here - https://github.com/eslint/eslint/issues/8229

**Please describe what the rule should do:**
`no-restricted-globals` allows to prevent the use of certain globals, but it doesn't allow providing custom messages that would explain why the usage is being prevented.

**What category of rule is this? (place an "X" next to just one item)**

[x] Enforces code style
[x] Warns about a potential error
[x] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// should warn/error with an explanation on use of window and suggest usage of self instead
if (window.crypto && window.crypto.getRandomValues) {
    // Modern browsers
    Rand.prototype._rand = function _rand(n) {
      var arr = new Uint8Array(n);
      window.crypto.getRandomValues(arr);
      return arr;
    };
  } else if (window.msCrypto && window.msCrypto.getRandomValues) {
    // IE
    Rand.prototype._rand = function _rand(n) {
      var arr = new Uint8Array(n);
      window.msCrypto.getRandomValues(arr);
      return arr;
    };
  } else {
    // Old junk
    Rand.prototype._rand = function() {
      throw new Error('Not implemented yet');
    };
  }

===============

// should not warn/error on the usage of self
if (self.crypto && self.crypto.getRandomValues) {
    // Modern browsers
    Rand.prototype._rand = function _rand(n) {
      var arr = new Uint8Array(n);
      self.crypto.getRandomValues(arr);
      return arr;
    };
  } else if (self.msCrypto && self.msCrypto.getRandomValues) {
    // IE
    Rand.prototype._rand = function _rand(n) {
      var arr = new Uint8Array(n);
      self.msCrypto.getRandomValues(arr);
      return arr;
    };
  } else {
    // Old junk
    Rand.prototype._rand = function() {
      throw new Error('Not implemented yet');
    };
  }
```

**Comment by platinumazure** [comment]
Thanks for the rule suggestion! I'll champion this.

**Comment by dryajov** [comment]
Thanks @platinumazure!

**Comment by ilyavolodin** [comment]
@dryajov Could you provide a schema of what you think the options should look like with this change?

**Comment by not-an-aardvark** [comment]
@platinumazure This has three :+1:s from the team -- marking as accepted.

**Comment by platinumazure** [comment]
@ilyavolodin Missed your question about proposed schema for this rule. I would want to do something similar to what we have in `no-restricted-syntax`. Maybe something like this:

```js
        schema: {
            type: "array",
            items: [{
                oneOf: [
                    {
                        type: "string"
                    },
                    {
                        type: "object",
                        properties: {
                            global: { type: "string" },
                            message: { type: "string" }
                        },
                        required: ["global"],
                        additionalProperties: false
                    }
                ]
            }],
            uniqueItems: true,
            minItems: 0
        }
```

**Comment by platinumazure** [comment]
Working on this.
